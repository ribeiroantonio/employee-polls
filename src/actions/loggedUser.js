export const LOG_USER_IN = "LOG_USER_IN";

const logUserIn = (id) => {
  return {
    type: LOG_USER_IN,
    id,
  };
};

export default logUserIn;
