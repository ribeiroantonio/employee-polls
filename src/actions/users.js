export const RETRIEVE_USERS = "RETRIEVE_USERS";
export const ADD_ANSWER_TO_USER = "ADD_ANSWER_TO_USER";
export const ADD_QUESTION_TO_USER = "ADD_QUESTION_TO_USER";

export const retrieveUsers = (users) => {
  return {
    type: RETRIEVE_USERS,
    users,
  };
};

export const addAnswerToUser = ({ loggedUser, questionId, answer }) => {
  return {
    type: ADD_ANSWER_TO_USER,
    loggedUser,
    questionId,
    answer,
  };
};

export const addQuestionToUser = ({ id, author }) => {
  return {
    type: ADD_QUESTION_TO_USER,
    id,
    author,
  };
};
