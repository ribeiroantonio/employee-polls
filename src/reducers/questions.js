import {
  RETRIEVE_QUESTIONS,
  SAVE_QUESTION,
  SAVE_ANSWER,
} from "../actions/questions";

export default function questions(state = {}, action) {
  switch (action.type) {
    case RETRIEVE_QUESTIONS:
      return {
        ...state,
        ...action.questions,
      };
    case SAVE_QUESTION:
      return {
        ...state,
        ...questions,
        [action.question.id]: action.question,
      };
    case SAVE_ANSWER:
      const { loggedUser, questionId, answer } = action;
      return {
        ...state,
        [questionId]: {
          ...state[questionId],
          [answer]: {
            ...state[questionId][answer],
            votes: state[questionId][answer].votes.concat([loggedUser]),
          },
        },
      };
    default:
      return state;
  }
};
