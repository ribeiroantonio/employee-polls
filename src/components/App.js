import React, {useEffect} from 'react';
import { connect } from 'react-redux';
import {handleInitialData} from "../actions/shared";
import { Route, Routes } from 'react-router-dom';
import Nav from './Nav';
import Login from './Login'
import Leaderboard from './Leaderboard'
import NewPoll from './NewPoll'
import Poll from './Poll'
import Error from './Error'
import Dashboard from './Dashboard'

function App(props) {
  useEffect(() => {
    props.dispatch(handleInitialData());
  });

  return (
    <div>
      {props.userLoggedIn === false ? (<Login />) : (
        <div>
          <Nav/>
          <Routes>
            <Route path="/" exact element={<Dashboard/>} />
            <Route exact path="/login" element={<Login />} />
            <Route exact path="/leaderboard" element={<Leaderboard />} />
            <Route exact path="/add" element={<NewPoll />} />
            <Route path="/questions/:questionId" element={<Poll />} />
            <Route path="*" element={<Error />} />
          </Routes>
        </div>
      )}
    </div>
  );
}

const mapStateToProps = ({loggedUser}) => ({
  userLoggedIn: !!loggedUser,
});

export default connect(mapStateToProps)(App);
