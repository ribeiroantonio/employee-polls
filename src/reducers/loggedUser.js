import { LOG_USER_IN } from "../actions/loggedUser";

export default function loggedUser(state = null, action) {
  switch (action.type) {
    case LOG_USER_IN:
      return action.id;
    default:
      return state;
  }
};
