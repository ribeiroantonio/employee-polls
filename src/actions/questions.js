import { saveQuestion } from "../utils/api";
import { addQuestionToUser } from "./users";

export const RETRIEVE_QUESTIONS = "RETRIEVE_QUESTIONS";
export const SAVE_QUESTION = "SAVE_QUESTION";
export const SAVE_ANSWER = "SAVE_ANSWER";

export const retrieveQuestions = (questions) => {
  return {
    type: RETRIEVE_QUESTIONS,
    questions,
  };
};

export const saveUserQuestion = (question) => {
  return {
    type: SAVE_QUESTION,
    question,
  };
};

export const handleAddQuestion = (question) => {
  return (dispatch) => {
    return saveQuestion(question).then((question) => {
      dispatch(saveUserQuestion(question));
      dispatch(addQuestionToUser(question));
    });
  };
};

export const saveQuestionAnswer = ({ loggedUser, questionId, answer }) => {
  return {
    type: SAVE_ANSWER,
    loggedUser,
    questionId,
    answer,
  };
};
