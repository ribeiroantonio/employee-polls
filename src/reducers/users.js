import { RETRIEVE_USERS } from "../actions/users";
import { ADD_ANSWER_TO_USER } from "../actions/users";
import { ADD_QUESTION_TO_USER } from "../actions/users";

export default function users (state = {}, action) {
  switch (action.type) {
    case RETRIEVE_USERS:
      return {
        ...state,
        ...action.users,
      };
    case ADD_QUESTION_TO_USER:
      const { id, author } = action;
      return {
        ...state,
        [author]: {
          ...state[author],
          questions: state[author].questions.concat([id]),
        },
      };
    case ADD_ANSWER_TO_USER:
      const { loggedUser, questionId, answer } = action;
      return {
        ...state,
        [loggedUser]: {
          ...state[loggedUser],
          answers: {
            ...state[loggedUser].answers,
            [questionId]: answer,
          },
        },
      };
    default:
      return state;
  }
};
