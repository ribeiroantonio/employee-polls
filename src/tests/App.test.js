import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { configureStore } from '@reduxjs/toolkit';
import loggedUser from  '../reducers/loggedUser';
import questions from  '../reducers/questions';
import users from  '../reducers/users';
import App from "../components/App";
import logUserIn from "../actions/loggedUser";

const store = configureStore({
  reducer: {
    loggedUser,
    questions,
    users,
  },
});

describe("when rendering error page", () => {
  it("matches snapshot", () => {
    const user = "sarahedo"
    store.dispatch(logUserIn(user));

    const { container } = render(
      <MemoryRouter>
        <Provider store={store}>
          <App />
        </Provider>
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
