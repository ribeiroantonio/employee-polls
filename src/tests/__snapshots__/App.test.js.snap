// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when rendering error page matches snapshot 1`] = `
<div>
  <div>
    <div>
      <nav
        class="navbar navbar-expand-lg bg-light"
      >
        <div
          class="container-fluid"
        >
          <div
            class="collapse navbar-collapse"
            id="navbarSupportedContent"
          >
            <ul
              class="navbar-nav me-auto mb-2 mb-lg-0"
            >
              <li
                class="nav-item"
              >
                <a
                  aria-current="page"
                  class="nav-link active"
                  href="/"
                >
                  Home
                </a>
              </li>
              <li
                class="nav-item"
              >
                <a
                  class="nav-link active"
                  href="/leaderboard"
                >
                  Leaderboard
                </a>
              </li>
              <li
                class="nav-item"
              >
                <a
                  class="nav-link active"
                  href="/add"
                >
                  New
                </a>
              </li>
            </ul>
            <div
              class="m-2"
            >
              Logged in as: 
              sarahedo
            </div>
            <button
              class="btn btn-outline-success"
            >
              Logout
            </button>
          </div>
        </div>
      </nav>
      <div
        class="container mt-4"
      >
        <ul
          class="nav nav-tabs"
          id="pollsTab"
          role="tablist"
        >
          <li
            class="nav-item"
            role="presentation"
          >
            <button
              aria-controls="home-tab-pane"
              aria-selected="true"
              class="nav-link active"
              data-bs-target="#home-tab-pane"
              data-bs-toggle="tab"
              id="home-tab"
              role="tab"
              type="button"
            >
              New Polls: 
              0
            </button>
          </li>
          <li
            class="nav-item"
            role="presentation"
          >
            <button
              aria-controls="profile-tab-pane"
              aria-selected="false"
              class="nav-link"
              data-bs-target="#profile-tab-pane"
              data-bs-toggle="tab"
              id="profile-tab"
              role="tab"
              type="button"
            >
              Answered Polls: 
              0
            </button>
          </li>
        </ul>
        <div
          class="tab-content"
          id="myTabContent"
        >
          <div
            aria-labelledby="home-tab"
            class="tab-pane fade show active"
            id="home-tab-pane"
            role="tabpanel"
            tabindex="0"
          >
            <div
              class="card"
            >
              <div
                class="card-body"
              />
            </div>
          </div>
          <div
            aria-labelledby="profile-tab"
            class="tab-pane fade"
            id="profile-tab-pane"
            role="tabpanel"
            tabindex="0"
          >
            <div
              class="card"
            >
              <div
                class="card-body"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
